src/main/java/org/broadinstitute/hellbender/tools/exome/copyratio/CopyRatioSamplers.java

```diff
@@ -0,0 +1,165 @@
+package org.broadinstitute.hellbender.tools.exome.copyratio;
+
+import org.apache.commons.math3.distribution.BetaDistribution;
+import org.apache.commons.math3.random.RandomGenerator;
+import org.broadinstitute.hellbender.utils.MathUtils;
+import org.broadinstitute.hellbender.utils.mcmc.ParameterSampler;
+import org.broadinstitute.hellbender.utils.mcmc.SliceSampler;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+
+/**
+ * @author Samuel Lee &lt;slee@broadinstitute.org&gt;
+ */
+final class CopyRatioSamplers {
+    private CopyRatioSamplers() {}
+
+    //Calculates the exponent for a normal distribution; used in log-likelihood calculation below.
+    private static double normalTerm(final double quantity, final double mean, final double variance) {
+        return (quantity - mean) * (quantity - mean) / (2. * variance);
+    }
+
+    //samples log conditional posterior for the variance parameter, assuming uniform prior; this is given by
+    //the product of Gaussian likelihoods for each non-outlier target t:
+    //  log[product_{non-outlier t} variance^(-1/2) * exp(-(coverage_t - mean_t)^2 / (2 * variance))] + constant
+    //where mean_t is identical for all targets in a segment
+    protected static final class VarianceSampler implements ParameterSampler<Double, CopyRatioParameter, CopyRatioState, CopyRatioData> {
+        private final double varianceMin;
+        private final double varianceMax;
+        private final double varianceSliceSamplingWidth;
+
+        public VarianceSampler(final double varianceMin, final double varianceMax, final double varianceSliceSamplingWidth) {
+            this.varianceMin = varianceMin;
+            this.varianceMax = varianceMax;
+            this.varianceSliceSamplingWidth = varianceSliceSamplingWidth;
+        }
+
+        public Double sample(final RandomGenerator rng, final CopyRatioState state, final CopyRatioData dataCollection) {
+            final Function<Double, Double> logConditionalPDF = newVariance -> {
+                double ll = 0.;
+                int numNotOutliers = 0;
+                for (int segment = 0; segment < dataCollection.getNumSegments(); segment++) {
+                    final int numTargetsInSegment = dataCollection.getNumTargetsInSegment(segment);
+                    final int targetStartInSegment = dataCollection.getStartTargetInSegment(segment);
+                    for (int target = 0; target < numTargetsInSegment; target++) {
+                        final double coverage = dataCollection.getCoveragesInSegment(segment).get(target);
+                        if (!state.targetOutlierIndicator(targetStartInSegment + target)) {
+                            ll -= normalTerm(coverage, state.segmentMean(segment), newVariance);
+                            numNotOutliers++;
```

Comment:
Keeping track of `numNonOutliers` is messy.  Instead, before the loop define `gaussianLogNormalization = -0.5 * Math.log(newVariance)` and in the loop have ` ll -= normalTerm(. . .) + gaussianLogNormalization`.